### PR TITLE
fix(kb): pool async LanceDB connections across event loops (#2200)

### LINKED ISSUE #2200: fix(kb): store async-connection init deadlocks when reached from more than one event loop
## Why

`LanceDBVectorIndexStore._async_lock` (`storage/lancedb_stores.py:646`) and `LanceDBIngestionStatusStore._async_lock` (`:3641`) are `asyncio.Lock()` objects built in `__init__` to guard lazy `_async_conn` initialization. But both stores are **process-wide singletons** held by `StorageFactory` (`storage/factory.py:159/189/203/217/231`), while the codebase repeatedly creates **new event loops**: `kb/coordinator.py:2329/2332/2341` (2341 on a fresh thread), `management/collection_manager.py:175/202`, `pipelines/document_ingestion.py:1143`, `parse/parse_document.py:110`.

One `asyncio.Lock` reached from more than one event loop deadlocks. In CPython 3.12, `Lock.acquire` returns on the uncontended path by setting `self._locked = True` **without ever touching `_get_loop()`**, and `_LoopBoundMixin._get_loop` binds `self._loop` only on its first call. So: thread A takes the lock uncontended and `_loop` is still `None`; thread B contends, `_get_loop()` binds `_loop` to **loop B**, and B awaits a future created on loop B; when A releases, `_wake_up_first()` calls `fut.set_result(True)` directly, **not** through `call_soon_threadsafe`, so loop B is never woken and thread B hangs forever. With the opposite binding order it surfaces as `RuntimeError: <Lock ...> is bound to a different event loop` instead.

Minimal reproduction, mirroring the fast-path + double-check shape of `_get_async_connection`:

```python
import asyncio, threading
lock = asyncio.Lock()
state = {"conn": None}

async def get_async_connection():
    if state["conn"] is not None:
        return state["conn"]
    async with lock:
        if state["conn"] is not None:
            return state["conn"]
        await asyncio.sleep(0.3)          # stands in for lancedb.connect_async
        state["conn"] = object()
        return state["conn"]

for _ in range(2):
    threading.Thread(target=lambda: asyncio.run(get_async_connection())).start()
```

Observed: one thread completes, the other **hangs forever and raises nothing**.

**Impact**: the window is the first few concurrent KB requests after process start (its width is however long `lancedb.connect_async` takes); once `_async_conn` is initialized it can no longer be hit. A request that does hit it hangs forever or 500s.

## Scope

This is the other half of the synchronous-side problem. Sync connections already have a process-wide, `RLock`-guarded, TTL'd pool (`LanceDBConnectionManager` in `providers/vector_store/lancedb.py`), which is why #657 deletes five redundant per-instance sync caches. Async connections have **no** such pool, so each store caches `_async_conn` itself behind an `asyncio.Lock` — and that is where the deadlock comes from.

Prefer fixing it symmetrically rather than swapping in a `threading.Lock` per store:

- Add a process-wide async connection pool next to `LanceDBConnectionManager`, guarded by a `threading.Lock` (not an `asyncio.Lock`), keyed by URI.
- Delete `LanceDBVectorIndexStore._async_conn` / `_async_lock` and `LanceDBIngestionStatusStore._async_conn` / `_async_lock`; point both `_get_async_connection` methods at the pool.
- Net effect is mostly deleted code, and it matches how the sync side already works.

## Acceptance Criteria

- Concurrently triggering async connection init from two different event loops (including cross-thread `asyncio.run`) neither hangs nor raises `bound to a different event loop`.
- The reproduction above is pinned as a regression test that hangs/fails before the fix, with a timeout so CI cannot stall on it.
- Async methods stay async and sync methods stay sync (the #494 invariant).
- Return shapes and exception mapping are unchanged; no regression in the KB suites.

## Non-goals

- Do not change the shape of the public store contracts.
- Do not unify the sync and async pools' TTL policy or configuration while here.
- Do not remove the `asyncio.run` / `new_event_loop` call sites; multiple loops are pre-existing architecture, and this issue only makes the store safe for them.

Part of #494. Adjacent to #657 but non-overlapping: #657 covers the sync connection caches and the table-handle cache, this covers async connection initialization.

### PR DESCRIPTION
**Pull request by bluefish-08**
## Invariant

No asyncio synchronization primitive is stored on an object that outlives a single event loop.

## The bug

`LanceDBVectorIndexStore._async_lock` and `LanceDBIngestionStatusStore._async_lock` were `asyncio.Lock()` objects built in `__init__` to guard lazy `_async_conn` init. Both stores are process-wide singletons (`storage/factory.py`), while the codebase keeps creating new event loops — `kb/coordinator.py:2329/2332/2341` (one on a fresh thread), `management/collection_manager.py:175/202`, `pipelines/document_ingestion.py:1143`, `parse/parse_document.py:110`.

In CPython 3.12 `Lock.acquire` returns on the uncontended path without touching `_get_loop()`, and `_LoopBoundMixin._get_loop` binds on its first call. Thread A takes the lock uncontended, leaving `_loop` unset; thread B contends, binds `_loop` to loop B, and awaits a future on loop B; A's release calls `fut.set_result(True)` directly rather than through `call_soon_threadsafe`, so loop B is never woken. Thread B hangs forever, raising nothing. The window is the first few concurrent KB requests after process start.

## The fix

Sync connections already come from a process-wide, `RLock`-guarded pool (`LanceDBConnectionManager`). Async connections had no equivalent, so each store cached its own behind an `asyncio.Lock` — that is the whole bug. This adds the missing pool, guarded by a `threading.Lock`, and deletes the per-instance state:

- `get_async_connection_from_env()` in `providers/vector_store/lancedb.py`, double-checked fill with `connect_async` awaited outside the lock; the connection that loses the insert race is closed.
- Both stores lose `_async_conn` / `_async_lock`, and `LanceDBIngestionStatusStore.__init__` goes away entirely. Each `_get_async_connection` is now one line.
- `resolve_dir_from_env()` is factored out of `get_connection_from_env()` so both planes resolve the directory the same way. The async path uses it directly instead of opening a sync connection just to read `.uri` off it — that indirection made 14 existing async store tests silently hit the real filesystem, since they patch `lancedb_stores.get_connection_from_env` and the uri lookup had moved into the provider module. Measured: 11 real `lancedb.connect()` calls before this, 0 after, matching the pre-change baseline.

`clear_connection_cache()` now drops async entries too, because the pool outlives the store instances that `StorageFactory.reset_all()` used to throw away, and `storage_shim.reset_rag_storage_for_tests` promises to reset *all* process-global state. It deliberately does **not** close them: stores hold a connection across awaits, so closing here strands an in-flight coroutine with `RuntimeError: Connection is closed`, and dropping the entry is already enough to force a fresh connect.

## The premise, checked

The pool only works if an AsyncConnection is not bound to its creating loop. Verified against real lancedb 0.33.0 across loops that had already closed: reads, `create_table`, `open_table` + `add`, a table handle created on one loop and written from another, 8 reads + 4 writes gathered on one loop, 20 successive loops on one connection, and 4 loops live at once. All clean. The mechanism is that pyo3-async-runtimes calls `get_running_loop()` per method call and stores no loop on the object.

`test_real_async_connection_survives_a_second_event_loop` pins this with real lancedb, so a future upgrade that starts binding connections to loops fails here rather than silently. The other tests fake `connect_async`, which is why that one is worth its cost.

## Not included

- The `asyncio.run` / `new_event_loop` call sites stay; multiple loops are pre-existing architecture and this only makes the store safe for them.
- No TTL on the async pool. `LANCEDB_DIR` is process-level and nothing in `src/` rewrites it per request, so the pool holds one entry in production.
- **This is one instance of a wider bug class.** A sweep found 17 more process-level `asyncio` primitives reachable from more than one loop — most notably `RAG_tools/progress/realtime.py:31` (the global `progress_broadcaster`'s lock, already provably cross-loop: connect/disconnect on the uvicorn loop while ingestion's temporary loop does `create_task(broadcast_progress)`) and six in `web/sandbox_manager.py` (a singleton with no lifespan-rebuild at all). `web/api/websocket.py:4353` already fixes this exact class for one lock, and `web/tools/config.py:601` / `collection_manager.py:267` show the loop-keyed pattern. Filing those separately rather than widening this PR.

## Change breakdown

Source `+62/-40` across two files, plus a 218-line test module. Tests are 78% of the diff.

## Validation

- Green: `RAG_tools` storage, kb, version_management, retrieval, management, LanceDB, plus `providers/vector_store/test_lancedb.py`. The one failure in `pipelines` is pre-existing and environmental (local modelhub sqlite lacks `models.managed_by`).
- `pre-commit run --files` clean across every hook.
- New tests: 8 consecutive runs, 0 failures. `JOIN_TIMEOUT` is 5s against a 0.2s happy path, so a regression fails fast instead of parking CI.

### Mutation matrix

| Mutation | Detected |
|---|---|
| Restore per-instance `_async_conn` + `asyncio.Lock` (the bug) | 5/5 — `a store's async connection init never returned -- deadlock`, 34s of join timeout |
| `clear_connection_cache` skips the async pool | 5/5 |
| Drop the close of the connection that lost the race | 5/5 |
| Pool stops caching | 5/5, all four fake-backed tests |

Closes #2200

### CODE REVIEW

File: tests/core/tools/core/RAG_tools/storage/test_lancedb_stores_concurrency.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
+"""The store data plane must survive concurrent handle access.
+
+The coordinator offloads blocking handle calls with ``asyncio.to_thread``. Once
+those are awaited concurrently on a shared loop, several worker threads reach a
+single process-wide store instance at the same time. These tests pin the two
+pieces of *synchronous* shared state that made that unsafe: the table-handle
+cache, and the per-instance sync connection cache that no longer exists.
+
+Async connection init is NOT covered here and is not safe yet: ``_async_conn``
+is still guarded by an ``asyncio.Lock`` that deadlocks when reached from more
```
Comment (rogercloud) [inline tests/core/tools/core/RAG_tools/storage/test_lancedb_stores_concurrency.py]: Stale comment: says async connection init "is NOT covered here and is not safe yet" via `_async_conn`/`asyncio.Lock`, and references #2200. This PR closes #2200 and deletes `_async_conn`/`_async_lock` entirely — please update or remove this comment.

File: tests/providers/vector_store/test_lancedb_async_pool.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            results.append(asyncio.run(coro_factory()))
+        except BaseException as exc:  # noqa: BLE001
+            errors.append(f"{type(exc).__name__}: {exc}")
+
+    thread = threading.Thread(target=target, daemon=True)
+    thread.start()
+    return thread
+
+
+def test_concurrent_init_from_two_loops_does_not_deadlock() -> None:
```
Comment (rogercloud) [inline tests/providers/vector_store/test_lancedb_async_pool.py:74]: All tests here (lines 74, 99, 117, 148, 179) exercise cross-thread/cross-loop concurrency only. Consider adding one same-event-loop concurrent test (e.g. via `asyncio.gather`) to pin the no-await-under-lock invariant, so a future regression that moves an await inside the locked section is caught.

File: tests/providers/vector_store/test_lancedb_async_pool.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    """Two threads, two loops, one uninitialized pool entry: both must return."""
+    connect_async, created = _slow_connect_async()
```
Comment (rogercloud) [inline tests/providers/vector_store/test_lancedb_async_pool.py:76]: `test_concurrent_init_from_two_loops_does_not_deadlock` relies on a 0.2s sleep to force the race but never asserts `len(created) == 2` (or equivalent). Without that assertion, the test can silently degenerate into testing plain sequential reuse if the two callers don't overlap. Add an assertion on the number of `connect_async` calls.

File: tests/providers/vector_store/test_lancedb_async_pool.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
+"""Async LanceDB connections must be reachable from every event loop.
```
Comment (rogercloud) [inline tests/providers/vector_store/test_lancedb_async_pool.py:1]: No test simulates `lancedb.connect_async` raising an exception. The implementation looks correct by construction (pool dict only written after success), but add a test asserting the exception propagates cleanly and the pool stays untouched.

File: src/xagent/providers/vector_store/lancedb.py [hunk 5]
```diff
@@ -37,18 +38,64 @@
...
+        _async_connection_cache.clear()
+
+
+async def get_async_connection_from_env(env_var: str = "LANCEDB_DIR") -> Any:
+    """Get a process-wide async LanceDB connection for the env-configured dir.
+
+    Both planes resolve the directory through ``resolve_dir_from_env``, so they
+    always agree on it without this having to open a sync connection first.
+    """
+    uri = LanceDBConnectionManager().resolve_dir_from_env(env_var)
```
Comment (rogercloud) [inline src/xagent/providers/vector_store/lancedb.py:75]: The directory-parity fix (unifying async on `resolve_dir_from_env()`, matching sync at line 249) has no direct test for the unset/empty `LANCEDB_DIR` case — exactly where the old sync/async divergence lived. All existing pool tests set `LANCEDB_DIR` explicitly. Add a test covering unset/empty env and sync/async parity.

File: src/xagent/providers/vector_store/lancedb.py [hunk 6]
```diff
@@ -37,18 +38,64 @@
...
 
 
 def clear_connection_cache() -> None:
-    """Clear the global LanceDB connection cache.
+    """Clear the global LanceDB connection caches, sync and async.
 
     This is primarily intended for test isolation to avoid reusing cached
     connections across different `LANCEDB_DIR` values.
+
+    Both halves only drop their entries. Closing a pooled async connection here
```
Comment (rogercloud) [inline src/xagent/providers/vector_store/lancedb.py:59]: `clear_connection_cache()` drops async pool entries via `.clear()` without closing them, relying on refcounting/Rust `Drop`. This is fine for a fixed `LANCEDB_DIR` per process, but suites that rotate `LANCEDB_DIR` per test (e.g. `isolate_rag_storage`) accumulate unclosed connections. Consider an explicit close as a follow-up.

File: src/xagent/core/tools/core/RAG_tools/storage/lancedb_stores.py [hunk 7]
```diff
@@ -3633,26 +3625,13 @@ class LanceDBIngestionStatusStore(IngestionStatusStore):
...
-        self._async_lock = asyncio.Lock()
-
     def _get_sync_connection(self) -> DBConnection:
         """Get sync LanceDB connection."""
-        if self._sync_conn is None:
-            self._sync_conn = get_connection_from_env()
-        return self._sync_conn
+        return get_connection_from_env()
 
     async def _get_async_connection(self) -> Any:
```
Comment (rogercloud) [inline src/xagent/core/tools/core/RAG_tools/storage/lancedb_stores.py]: `_get_async_connection` has zero production callers — all `*_async` methods on this class delegate to sync implementations. Only the new test file calls it. Consider removing this dead method rather than rewiring it to the new pool.

**Review by rogercloud — APPROVED** [review]
### Minor

- `tests/core/tools/core/RAG_tools/storage/test_lancedb_stores_concurrency.py:9-11` — comment still says async init "is NOT covered" and deadlocks via `_async_conn`/`asyncio.Lock`; both are deleted by this PR. Update or remove the stale comment.
- `tests/providers/vector_store/test_lancedb_async_pool.py:74,99,117,148,179` — all cases test cross-thread/cross-loop concurrency only; no test exercises two coroutines on the same event loop (e.g. via `asyncio.gather`). Add one same-loop test to pin the no-await-under-lock invariant against future regressions.
- `tests/providers/vector_store/test_lancedb_async_pool.py:74-96` — `test_concurrent_init_from_two_loops_does_not_deadlock` relies on a 0.2s sleep to force the race but never asserts `len(created) == 2`. Add that assertion so the test fails loudly if the race stops overlapping.
- `tests/providers/vector_store/test_lancedb_async_pool.py` — no test simulates `connect_async` raising. Add a case asserting the pool dict stays untouched and the exception propagates for clean retry.
- `src/xagent/providers/vector_store/lancedb.py:75,249` — the directory-parity fix (unifying sync/async on `resolve_dir_from_env()`) has no test for the unset/empty `LANCEDB_DIR` path, which is exactly where the original divergence lived. Add a direct test for `resolve_dir_from_env()` covering unset/empty env and a sync/async parity check.
- `src/xagent/providers/vector_store/lancedb.py:53-66` — `clear_connection_cache()` drops async pool entries without closing them, relying on refcounting/Rust `Drop`. Fine for the documented single-directory-per-process case, but test suites rotating `LANCEDB_DIR` per test (e.g. `isolate_rag_storage`) accumulate unclosed connections. Consider an explicit close on cache clear as a follow-up.
- `src/xagent/core/tools/core/RAG_tools/storage/lancedb_stores.py:3632` — `LanceDBIngestionStatusStore._get_async_connection` has no production callers (all `*_async` methods delegate to sync); only the new test file calls it. Remove it or wire it up, rather than rewiring dead code to the new pool.

Blocking: no — recommended event: APPROVE

**Comment by bluefish-08** [comment]
Rebased onto main now that #2201 has landed. This is down to its own two commits, `+287/-39`.

One conflict came up and is worth naming: #2201 later added `await asyncio.to_thread(self._get_connection)` inside the old `_get_async_connection`, while this PR deletes that method body outright in favour of the pool. I took this PR's side — the method is now one line — so that `to_thread` change is subsumed rather than lost. `LanceDBVectorIndexStore.__init__` merged cleanly with both sides intact: `_table_cache_epoch` from #2201, and no `_async_conn`/`_async_lock` from here.

Re-verified after the rebase: RAG_tools storage, kb, version_management, retrieval, management, LanceDB, plus both providers/vector_store suites — all green; `pre-commit run --files` clean.

File: tests/core/tools/core/RAG_tools/storage/test_lancedb_stores_concurrency.py (hunk 1 above)
Comment (bluefish-08) [reply tests/core/tools/core/RAG_tools/storage/test_lancedb_stores_concurrency.py]: Fixed in f9f65de3 — it now points at `test_lancedb_async_pool.py` and describes the pool as the resolution of #2200 rather than an open gap.

File: tests/providers/vector_store/test_lancedb_async_pool.py (hunk 2 above)
Comment (bluefish-08) [reply tests/providers/vector_store/test_lancedb_async_pool.py:74]: Added, but not the way you suggested, because I tried that first and it does not work.

An `asyncio.gather` test cannot catch an await moved under `_async_cache_lock`. I mutated the code that way to check: the lock is a *threading* lock, so the first coroutine holding it across an await blocks the whole thread, which stops the event loop, which means `asyncio.wait_for` never gets to fire. The suite hangs instead of failing — my run had to be killed after 10 minutes. That is worse than no test, since CI would sit there until its own timeout.

So the invariant is now pinned statically: `test_no_await_inside_the_pool_lock` parses `get_async_connection_from_env` and fails if any `Await`/`AsyncWith`/`AsyncFor` appears under a `with _async_cache_lock:` block. Same mutation is caught instantly. `test_coordinator_router_boundary.py` set the precedent for AST guards here.

I kept a same-loop `asyncio.gather` test too, but only asserting the behaviour it can actually assert: five gathered callers converge on one pooled connection and every superseded one is closed.

File: tests/providers/vector_store/test_lancedb_async_pool.py (hunk 3 above)
Comment (bluefish-08) [reply tests/providers/vector_store/test_lancedb_async_pool.py:76]: You are right, it could degenerate silently. Added `assert len(created) == 2` with a message naming the actual count. Verified by mutation: joining the two threads sequentially instead of concurrently now fails with `the two callers did not overlap (1 connect_async calls)`, where before it passed.

File: tests/providers/vector_store/test_lancedb_async_pool.py (hunk 4 above)
Comment (bluefish-08) [reply tests/providers/vector_store/test_lancedb_async_pool.py:1]: `test_failed_connect_leaves_the_pool_untouched` added: the exception propagates unchanged, the pool stays empty, and a subsequent call still succeeds. Mutation-checked by writing the failed uri into the cache before re-raising — the test fails.

File: src/xagent/providers/vector_store/lancedb.py (hunk 5 above)
Comment (bluefish-08) [reply src/xagent/providers/vector_store/lancedb.py:75]: Added, parametrized over unset / empty / whitespace-only `LANCEDB_DIR`. It asserts `resolve_dir_from_env()` matches the normalized `get_default_lancedb_dir()`, that the empty case raises, and that the async pool keys on exactly the directory the sync plane resolves. Mutation-checked by restoring the old async-side `os.getenv(env_var, "./data/lancedb")` — the unset case fails, which is precisely the divergence you pointed at.

File: src/xagent/providers/vector_store/lancedb.py (hunk 6 above)
Comment (bluefish-08) [reply src/xagent/providers/vector_store/lancedb.py:59]: Deliberate, and I agree it deserves a follow-up rather than a change here.

The reason it only clears: the stores hold a connection across awaits (`async_conn = await self._get_async_connection()` then `await async_conn.open_table(...)`), so closing under them strands an in-flight coroutine with `RuntimeError: Connection is closed`. I had it closing at first and backed it out after reproducing exactly that — and `tests/conftest.py` calls `clear_connection_cache()` in an autouse fixture around every test, so it would have been a live flake surface.

Your accumulation point stands for suites that rotate `LANCEDB_DIR`. Doing it safely needs refcounting or a close-on-idle path, which is more than this PR should carry. Happy to file it.

File: src/xagent/core/tools/core/RAG_tools/storage/lancedb_stores.py (hunk 7 above)
Comment (bluefish-08) [reply src/xagent/core/tools/core/RAG_tools/storage/lancedb_stores.py]: Confirmed and deleted in f9f65de3. All four `*_async` methods on `LanceDBIngestionStatusStore` say "Delegate to sync implementation for now" and none of them touch it — I was migrating a dead method onto the new pool. Good catch.

The test that used it now exercises two `LanceDBVectorIndexStore` instances instead, which still covers the point it was making (separate store objects, separate loops, one pooled connection). `LanceDBVectorIndexStore._get_async_connection` stays — that one has seven real callers.

**Comment by bluefish-08** [comment]
Filed the two follow-ups promised here: #2331 covers the wider cross-loop `asyncio`-primitive class (17 process-level singletons, 28 primitives, with `progress_broadcaster` as the proven case), and #2332 covers `clear_connection_cache` dropping sync connections without closing them.
